### ISSUE TITLE
Splash screens and cordova-plugin-xapkreader: Conflict?

### ISSUE THREAD

**Issue by lutzissler** [agamemnus/cordova-plugin-xapkreader#16]
First, thank you for your hard working making this great plugin available to the public. We definitely consider donating!

After adding cordova-plugin-xapkreader to a project (using Cordova 4.1.2), the splash screen defined in `config.xml` with

```
    <preference name="SplashScreen" value="screen" />
```

don’t show up anymore. To verify it is cordova-plugin-xapkreader causing the splash screen to not show up, I removed cordova-plugin-xapkreader from the project. In the project without cordova-plugin-xapkreader, the splash screen shows up as expected. Then, I re-added cordova-plugin-xapkreader, and the splash screen is gone again.

Debugging `CordovaActivity.loadUrl()`, which is responsible for finding the splash screen files, didn’t suggest anything going wrong. `CordovaActivity.showSplashScreen()` also worked smoothly.

Any thoughts how to fix this?

**Comment by agamemnus** [comment]
Hard to say. https://github.com/apache/cordova-plugin-splashscreen/blob/master/src/android/SplashScreen.java is the code.

Add a debug line inside onMessage:
Log.d (LOG_TAG, "135: " + id + " | " + data.toString());

Might also be something inside showSplashScreen.

Make a sample project with the splash screen and a dummy expansion file set up.

**Comment by lutzissler** [comment]
Sorry for the confusion with the version numbers. I was referring to the built-in splash screen in cordova-android 3.6.4. For that version of cordova-android, these are the steps to reproduce the error:

1. `cd` to a sandbox directory, like `cd /tmp`
2. `cordova create test test`
3. `cd test`
4. `cordova platform add android@3.6.4`
5. Create a red splash screen dummy: `convert -size 100x100 xc:red splash.png`
6. Put the dummy splash screen into the right folders: `cordova-splash` (using https://github.com/AlexDisler/cordova-splash)
7. Run the app: `cordova run --device` and see the splash screen appear
8. Add cordova-plugin-xapkreader: `cordova plugin add https://github.com/agamemnus/cordova-plugin-xapkreader.git`
9. Adjust `project.properties`
10. `platforms/android/cordova/clean` to wipe `ant-build` and `ant-gen`
11. Run the app: `cordova run --device` and see the splash screen **not appear this time**

Removing cordova-plugin-xapkreader from the project brings the splash screen back.

The code for `showSplashScreen()` on 3.6.4 is here: https://github.com/apache/cordova-android/blob/3.6.x/framework/src/org/apache/cordova/CordovaActivity.java#L931

The code for `loadUrl()` is here: https://github.com/apache/cordova-android/blob/3.6.x/framework/src/org/apache/cordova/CordovaActivity.java#L357

As mentioned initially, adding debug statements like the one you suggested to both methods doesn’t reveal anything suspicious. Any thoughts?

**Comment by agamemnus** [comment]
I'll test your steps in a few hours, after I wake up.

So the debug statements say that the ``showSplashScreen`` and ``loadUrl`` functions do run, even though you don't see the splash screen?

**Comment by lutzissler** [comment]
Correct, the debug statements indicate that the two function do run indeed. Furthermore, all the variables are containing values that make sense, like an ID for the splashscreen resource, for instance. Only I don’t see the splash screen.

**Comment by agamemnus** [comment]
Maybe it has something to do with the z-index of the view?

**Comment by lutzissler** [comment]
That was one of my thoughts, too, but on the other hand, it’s the `Dialog` that’s used to show the splash screen, so that shouldn’t be overlaid by anything else.

**Comment by agamemnus** [comment]
You forgot a step (the ``<preference name="SplashScreen" value="screen" />`` in config.xml)

We can rule out that there is something which overrides the preference, I guess.

Go through ``showSplashScreen`` step by step.
* Perhaps the app can't find the image? Check your logs carefully for any such errors or warnings.
* Maybe the width/height (e.g.: ``root.setMinimumHeight(display.getHeight());``) is somehow not working right... try checking what it is.
* Basically try to find any behavior difference that you can get the debug output for between one case and the other.

**Comment by lutzissler** [comment]
Comparing log outputs got me further. I have inserted log statements preceded with `Splash: ` at relevant lines in the code. In the debug outputs below, I have added line numbers for better reference.

This is the debug output *without* cordova-plugin-xapkreader:

	A01 D/CordovaActivity( 3695): CordovaActivity.init()
	A02 D/CordovaWebView( 3695): CordovaWebView is running on device made by: LGE
	A03 D/CordovaActivity( 3695): CordovaActivity.createViews()
	A04 D/CordovaActivity( 3695): Splash: Inside removeSplashScreen()
	A05 D/CordovaActivity( 3695): Splash: display.getHeight() = 1776
	A06 D/CordovaActivity( 3695): Splash: display.getWidth() = 1080
	A07 D/CordovaActivity( 3695): Splash: that.splashscreen = 2130837505
	A08 D/CordovaWebView( 3695): >>> loadUrl(file:///android_asset/www/index.html)
	A09 D/CordovaWebView( 3695): >>> loadUrlNow()
	A10 D/CordovaActivity( 3695): Resuming the App
	A11 D/CordovaWebViewClient( 3695): onPageStarted(file:///android_asset/www/index.html)
	A12 D/CordovaActivity( 3695): onMessage(onPageStarted,file:///android_asset/www/index.html)
	A13 D/CordovaLog( 3695): file:///android_asset/www/index.html: Line 26 : The key "target-densitydpi" is not supported.
	A14 D/CordovaWebViewClient( 3695): onPageFinished(file:///android_asset/www/index.html)
	A15 D/CordovaActivity( 3695): onMessage(onPageFinished,file:///android_asset/www/index.html)
	A16 D/CordovaActivity( 3695): onMessage(spinner,stop)
	A17 D/CordovaLog( 3695): file:///android_asset/www/js/index.js: Line 47 : Received Event: deviceready
	A18 D/CordovaActivity( 3695): onMessage(spinner,stop)
	A19 D/CordovaActivity( 3695): Splash: Inside removeSplashScreen()
	A20 D/CordovaActivity( 3695): Splash: splashDialog.dismiss()

And this is the debug output *with* cordova-plugin-xapkreader:

	B01 D/CordovaActivity( 3243): CordovaActivity.init()
	B02 D/CordovaWebView( 3243): CordovaWebView is running on device made by: LGE
	B03 D/CordovaActivity( 3243): CordovaActivity.createViews()
	B04 D/CordovaActivity( 3243): Splash: Inside removeSplashScreen()
	B05 D/CordovaActivity( 3243): Splash: display.getHeight() = 1776
	B06 D/CordovaActivity( 3243): Splash: display.getWidth() = 1080
	B07 D/CordovaActivity( 3243): Splash: that.splashscreen = 2130837506
	B08 D/CordovaWebView( 3243): >>> loadUrl(file:///android_asset/www/index.html)
	B09 D/CordovaWebView( 3243): >>> loadUrlNow()
	B10 D/CordovaActivity( 3243): Resuming the App
	B11 D/CordovaActivity( 3243): Paused the application!
	B12 D/CordovaWebView( 3243): Handle the pause
	B13 D/CordovaActivity( 3243): Splash: Inside removeSplashScreen()
	B14 D/CordovaActivity( 3243): Splash: splashDialog.dismiss()
	B15 D/CordovaActivity( 3243): Resuming the App
	B16 D/CordovaWebViewClient( 3243): onPageStarted(file:///android_asset/www/index.html)
	B17 D/CordovaActivity( 3243): onMessage(onPageStarted,file:///android_asset/www/index.html)
	B18 D/CordovaLog( 3243): : Line 1 : exception firing pause event from native
	B19 D/CordovaLog( 3243): : Line 1 : exception firing resume event from native
	B20 D/CordovaLog( 3243): file:///android_asset/www/index.html: Line 26 : The key "target-densitydpi" is not supported.
	B21 D/CordovaActivity( 3243): onMessage(spinner,stop)
	B22 D/CordovaWebViewClient( 3243): onPageFinished(file:///android_asset/www/index.html)
	B23 D/CordovaActivity( 3243): onMessage(onPageFinished,file:///android_asset/www/index.html)
	B24 D/CordovaLog( 3243): file:///android_asset/www/js/index.js: Line 47 : Received Event: deviceready
	B25 D/CordovaActivity( 3243): Splash: Inside removeSplashScreen()

Lines `A/B01`–`A/B10` are all good. If you look at `B11`, though, Cordova seems to detect an application pause. In response to the pause, Cordova hides the splash screen, which you can see in `CordovaActivity:590` (https://github.com/apache/cordova-android/blob/3.6.x/framework/src/org/apache/cordova/CordovaActivity.java#L590). According to the source comment, this is “to avoid leaking a window”. Cordova calls `removeSplashScreen()` which shows in lines `B13` and `B14` above. Lines `A11`–`A19` and `B16`, `B17`, `B20`–`B25` are equivalent again. Only that in the code *with* cordova-plugin-xapkreader `splashDialog.dismiss()` doesn’t get called in the end because the splash screen is already dismissed due to the pause.

I suspect that pause is caused by cordova-plugin-xapkreader, either directly or indirectly while reading the expansion file. Maybe Android’s taking over for the fraction of a second. Based on that suspicion, it looks like there are three solutions:

1. Prevent the pause to happen.
2. Make sure the pause happens before `showSplashScreen()` or after `removeSplashScreen()`.
3. Fix whatever “leaking a window” is, and then remove line 590 from `CordovaActivity.java`.

**Comment by agamemnus** [comment]
I think that (1) is the best solution here.. not sure how you can achieve (2) without a very deep knowledge of lots of code. (3) sounds potentially dangerous but I am not sure.

Best approach here, if you still want to use this splash screen stuff, would be try to find where the pause is happening inside the reader somehow...

**Comment by lutzissler** [comment]
Agreed, and of course I tried to find where it happens, but without luck. Frankly, I was somehow expecting that this would ring a bell on your end, which apparently it didn’t.

So before I waste loads of time step debugging all of your plugin's code, do you have at least a rough guess what to look at first in the reader?

**Comment by agamemnus** [comment]
Well, to be honest, I really don't know what kind of things can send the message about a 
"paused" application in that way. Maybe different threads? Who knows...

Here is a brief description of the control flow of this plugin:

1) The entry point is ``XAPKReader.java``: https://github.com/agamemnus/cordova-plugin-xapkreader/blob/gh-pages/src/android/XAPKReader.java. Don't think it's there.

2) That sets some variables including calling ``XAPKProvider.java`` via the .call function: https://github.com/agamemnus/cordova-plugin-xapkreader/blob/gh-pages/src/android/XAPKProvider.java#L83. Don't think it's there.

3) Then it runs an intent (``new Intent(context, XAPKDownloaderActivity.class);``) and starts ``XAPKDownloaderActivity.java``: https://github.com/agamemnus/cordova-plugin-xapkreader/blob/gh-pages/src/android/XAPKDownloaderActivity.java. Not sure if it's there.

Try commenting out these lines and see if that helps. It's a workaround for a Crosswalk bug that's fixed in newer versions: https://github.com/agamemnus/cordova-plugin-xapkreader/blob/gh-pages/src/android/XAPKDownloaderActivity.java#L65-L68.. I kind of have a brute force method of setting the right flags. Maybe clearFlags sets off the problem and it can be done in one command.

**Comment by lutzissler** [comment]
Thanks for the overview.

It’s this line: https://github.com/agamemnus/cordova-plugin-xapkreader/blob/gh-pages/src/android/XAPKReader.java#L60 So starting the activity/running the event causes the pause. Here’s some more debug output:

	D/XAPKReader(19304): startActivity
	I/ActivityManager(  745): START u0 {cmp=net.systemantics.app.test.xapkreader_splash/com.flyingsoftgames.xapkreader.XAPKDownloaderActivity (has extras)} from pid 19304
	D/CordovaWebView(19304): >>> loadUrlNow()
	D/CordovaActivity(19304): Resuming the App
	D/dalvikvm( 5784): GC_CONCURRENT freed 452K, 3% free 18455K/18976K, paused 2ms+1ms, total 25ms
	D/OpenGLRenderer(19304): Enabling debug mode 0
	D/dalvikvm( 2534): GC_CONCURRENT freed 402K, 4% free 17706K/18332K, paused 2ms+1ms, total 46ms
	D/SoftKeyboardDetect(19304): Ignore this event
	D/CordovaActivity(19304): Paused the application!
	D/CordovaWebView(19304): Handle the pause
	D/CordovaActivity(19304): Splash: Inside removeSplashScreen()
	D/CordovaActivity(19304): Splash: splashDialog.dismiss()
	W/InputMethodManagerService(  745): Focus gain on non-focused client com.android.internal.view.IInputMethodClient$Stub$Proxy@42b0db30 (uid=10074 pid=19304)
	D/XAPKDownloader(19304): Inside onCreate()

Is there any chance to start the activity before Cordova’s `loadUrl()` gets called (see https://github.com/apache/cordova-android/blob/3.6.x/framework/src/org/apache/cordova/CordovaActivity.java#L357), as that is setting up the splash screen?

**Comment by agamemnus** [comment]
I don't think so. But I am not entirely sure where/when loadUrl is called from my code.

I did some digging and it looks like ``loadUrl`` is called via ``execute()`` in ``CordovaLib/src/org/apache/cordova/App.java`` and also via ``showWebPage()`` in ``InAppBrowser.java``.

**Comment by lutzissler** [comment]
I think I found a fix … simply delay the activity start by the splash screen duration. It feels slightly hacky but I cannot see any possible interferences. Should also work with Cordova 4.x.x and the splash screen plugin used there. Just issued a pull request.

**Comment by agamemnus** [comment]
Ok.
2 issues:
1) There shouldn't be a delay by default... should be 0.
2) Needs better and consistent namespacing. So, ``xapk_expansion_load_delay`` instead would be better here.

**Comment by lutzissler** [comment]
The code needs to mimic the behavior from the core. Therefore, it has to use the same config parameter names, and also use the same default splash screen delay.

See here for the default 3000 ms timeout: https://github.com/apache/cordova-android/blob/3.6.x/framework/src/org/apache/cordova/CordovaActivity.java#L121

See here for getting `SplashScreenDelay` from the config: https://github.com/apache/cordova-android/blob/3.6.x/framework/src/org/apache/cordova/CordovaActivity.java#L361

But there’s one true point – the original code checks for the existence of the actual splash screen file, see https://github.com/apache/cordova-android/blob/3.6.x/framework/src/org/apache/cordova/CordovaActivity.java#L362-L366. If the file cannot be loaded (or is not specified), the splash screen delay shouldn’t be applied. So my fix should include lines 362–366 from `CordovaActivity.java`, too. I’ll look into that tomorrow and amend the pull request.

**Comment by agamemnus** [comment]
Oh, I see. I'm still not comfortable with this, not knowing the side-effects it could bring of the system loading in files (a second time) with the splash screen enabled. Let's have an ``xapk_expansion_ignore_splash_screen_delay`` variable (set to 0 by default) just in case.

**Comment by agamemnus** [comment]
Yo, any progress on this?

**Comment by lutzissler** [comment]
Well – as I was in the need of a workaround and the one I posted here as a pull request works, I didn’t invest more in that. To improve it, I think there’s two points needed:

1. Take these lines https://github.com/apache/cordova-android/blob/3.6.x/framework/src/org/apache/cordova/CordovaActivity.java#L362-L366 and apply them to my pull request after line 57, the `if` wrapping lines 59–68.

2. Add the config variable `xapk_expansion_ignore_splash_screen_delay` you had suggested above if you feel that’s necessary.

**Comment by agamemnus** [comment]
Okay, thanks. I'll make the changes soon-ish.

**Comment by agamemnus** [comment]
I am going to integrate the changes in the 5.3.1 version. Thanks.

**Comment by agamemnus** [comment]
Ok, so the latest code -- 4.1 -- doesn't even have a splash screen in the core code. I think they refactored the code. Did you make changes for 4.1, and if so, which changes?

**Comment by unskaa** [comment]
Oh man I spent a lot of time figuring out why my splash is not working on android anymore... I suppose I'll use lutz's fix.

**Comment by agamemnus** [comment]
Do splash screens load faster than if you just added an image to your app?

**Comment by unskaa** [comment]
In theory? Probably. I'm not expert on the matter but I suppose the splash screen is loaded before any DOM elements are being downloaded.

In practice? Probably in the millisecond range.

I find splash screen best because I can automatically prevent users from doing anything until deviceready.

**Comment by webplease** [comment]
I'm same problem/conflict with splashscreen, that not appears.

**Comment by agamemnus** [comment]
It is really a problem with the splash screen plugin running at the same time as this plugin. The splash screen plugin should go first but even if there was a hook to tell me that it was done, this would be adding a dependency to code that I would rather not maintain.

I can add a timer delay to my plugin at some point (default at 0), which the user should make sure to manually synchronize to the splash screen plugin.

**Comment by webplease** [comment]
Uhm, your plugin as expansion file support is awasome! ...but I think that for all is best to have a native splashscreen (ios/android).

What kind of Js splashscreen do u mean?

**Comment by agamemnus** [comment]
Just showing an image and removing it after your data has loaded.

**Comment by webplease** [comment]
What a pity! :(

**Comment by agamemnus** [comment]
If you really need it in now, I can drop everything and add the optional delay code into the new version... but I would want something in return. Perhaps a cushy job in your new administration.

**Comment by webplease** [comment]
Ahahahha :)

**Comment by agamemnus** [comment]
Seriously though, I suppose I can add it sometime this week since it seems to be a recurring issue/request...

**Comment by webplease** [comment]
yes, it would be fantastic! :)

**Comment by agamemnus** [comment]
I haven't forgotten. To-do for today!

**Comment by webplease** [comment]
wonderful!!!

**Comment by agamemnus** [comment]
I'm having some trouble with the new Github desktop interface. I made a new branch to make changes, but I can't load it in the interface. :(

**Comment by agamemnus** [comment]
Okay, I made the change on the ``cordova-5.3.1-updating`` branch.

As a first step, you will need to set the running time of the splash screen to a new preference called ``xapk_expansion_initialization_delay``.

**Comment by webplease** [comment]
Do not compile

Error:(81, 14) error: illegal start of type
Error:(83, 13) error: illegal start of type
Error:(83, 18) error: ';' expected
Error:(83, 19) error: invalid method declaration; return type required
Error:(83, 38) error: <identifier> expected
Error:(83, 47) error: <identifier> expected
Error:(84, 10) error: ')' expected
Error:(85, 6) error: reached end of file while parsing
Error:Execution failed for task ':compileDebugJavaWithJavac'.
> Compilation failed; see the compiler error output for details.

**Comment by agamemnus** [comment]
I'll look again... Did you want it to start from a function call it is a
countdown enough?

**Comment by agamemnus** [comment]
There was a missing semicolon on line 81.

**Comment by webplease** [comment]
right! :) now have published my app on Gplay ... after download, if the phone is not in Wifi the "downloading assetts" screen remains and not advance. When turn on wifi the "downloading assetts" process remains on 0/100 ...why ?

**Comment by agamemnus** [comment]
Not sure. Can you check if this happened with the stable/previous 5.3
version? (That is, one that doesn't count for this timer)

**Comment by webplease** [comment]
i'm on 5.3.1 ...uninstallig from Gplay e re-installing (in 4G too) the problem solved ... it's strange.

**Comment by agamemnus** [comment]
Well, it needs a clean install from Google Play, rather than just an update, I think.

Is the timer sufficient for you or do you need a function?

**Comment by webplease** [comment]
yes, delay works well! :) thanks very much

**Comment by cbacko01** [comment]
@agamemnus Thanks so much for working on the time-out work-around. 

I think a related issue is that the statusbar disappears when the XAPK plugin runs at startup. I delayed it until after my splash screen was done displaying, at which point, the statusbar disappeared. I was able to use the JS API and a timer to show the statusbar shortly thereafter, but it's a little distracting how the statusbar vanishes and comes back (causing the page to resize, too).

**Comment by agamemnus** [comment]
Actually that may be a result of my code here which is a workaround for a bug in Cordova/Crosswalk that has since been fixed in later versions. See if commenting that out fixes anything for you.

https://github.com/agamemnus/cordova-plugin-xapkreader/blob/cordova-5.3.1/src/android/XAPKDownloaderActivity.java#L74-L77:
````
cordovaActivity.getWindow().clearFlags(WindowManager.LayoutParams.FLAG_FORCE_NOT_FULLSCREEN);
cordovaActivity.getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN, WindowManager.LayoutParams.FLAG_FULLSCREEN);
````

**Comment by agamemnus** [comment]
Fixed per https://github.com/agamemnus/cordova-plugin-xapkreader/issues/64. Closing.
